File: src/main/java/com/rdvdev2/timetravelmod/impl/ModNetworking.java
package com.rdvdev2.timetravelmod.impl;

import com.rdvdev2.timetravelmod.api.dimension.ITimeline;
import com.rdvdev2.timetravelmod.impl.common.timemachine.TimeMachineExecutor;
import com.rdvdev2.timetravelmod.impl.common.timemachine.TimeMachineManager;
import com.rdvdev2.timetravelmod.impl.common.timemachine.exception.TimeMachineExecutionException;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.registry.Registry;
import net.minecraft.util.registry.RegistryKey;
import net.minecraft.world.Heightmap;
import net.minecraft.world.World;

import java.util.Arrays;
import java.util.Objects;
import java.util.Optional;

// TODO: Separate packets
public class ModNetworking {

    public static final Identifier OPEN_TIME_MACHINE_GUI = Mod.identifier("open_time_machine_gui");
    public static final Identifier OPEN_CREATIVE_TIME_MACHINE_GUI = Mod.identifier("open_creative_time_machine_gui");
    public static final Identifier OPEN_ENGINEER_BOOK_GUI = Mod.identifier("open_engineer_book_gui");

    public static final Identifier RUN_TIME_MACHINE = Mod.identifier("run_time_machine");
    public static final Identifier RUN_CREATIVE_TIME_MACHINE = Mod.identifier("run_creative_time_machine");

    public static void register() {
        ServerPlayNetworking.registerGlobalReceiver(RUN_TIME_MACHINE,
                (server, player, handler, buf, responseSender) -> {
                    Identifier worldIdentifier = buf.readIdentifier();
                    BlockPos rootPos = buf.readBlockPos();
                    Identifier timelineIdentifier = buf.readIdentifier();

                    server.execute(() -> {
                        World world = player.getServer().getWorld(RegistryKey.of(Registry.DIMENSION, worldIdentifier));
                        if (world == null) return;
                        ITimeline timeline = ModRegistries.TIMELINE.get(timelineIdentifier);
                        if (timeline == null) return;
                        Optional<TimeMachineExecutor> _tme = TimeMachineManager.getInstance().generateExecutor(world, rootPos);
                        _tme.ifPresent(tme -> {
                            try {
                                tme.checkAndRun(timeline, player, server);
                            } catch (TimeMachineExecutionException e) {
                                player.sendMessage(e.getError().getClientError(), true);
                                if (ModConfig.getInstance().getCommon().getEnableCheaterReports()) {
                                    Arrays.stream(server.getPlayerManager().getOpNames())
                                            .map(server.getPlayerManager()::getPlayer)
                                            .filter(p -> !Objects.equals(p, player))
                                            .forEach(p -> p.sendMessage(e.getError().getCheaterReport(player), false));
                                }
                            }
                        });
                    });
                });

        ServerPlayNetworking.registerGlobalReceiver(RUN_CREATIVE_TIME_MACHINE,
                (server, player, handler, buf, responseSender) -> {
                    Identifier timelineIdentifier = buf.readIdentifier();

                    server.execute(() -> {
                        ITimeline timeline = ModRegistries.TIMELINE.get(timelineIdentifier);
                        if (timeline == null) return;
                        ServerWorld world = server.getWorld(timeline.getWorld());
                        if (world == null || world.equals(player.world)) return;
                        player.getItemsHand().forEach( itemStack -> {
                            if (itemStack.getItem() == ModItems.CREATIVE_TIME_MACHINE) {
                                BlockPos pos = world.getTopPosition(Heightmap.Type.MOTION_BLOCKING_NO_LEAVES, player.getBlockPos());
                                if (pos.getY() == 0) pos = pos.up(world.getSeaLevel() + 2);
                                while (world.getBlockState(pos).isSolidBlock(world, pos) || world.getBlockState(pos.up()).isSolidBlock(world, pos.up())) pos = pos.up(); // Air space
                                while (!world.getBlockState(pos.down()).isSolidBlock(world, pos)) pos = pos.down(); // Touching ground
                                // TODO: Special logic for flying players
                                player.teleport(world, player.getX(), pos.getY(), player.getZ(), player.yaw, player.pitch);
                            }
                        });
                    });
                });
    }
}

File: src/main/kotlin/com/rdvdev2/timetravelmod/impl/ModNetworking.kt
package com.rdvdev2.timetravelmod.impl

import com.rdvdev2.timetravelmod.impl.Mod.identifier
import com.rdvdev2.timetravelmod.impl.ModConfig.instance
import com.rdvdev2.timetravelmod.impl.common.timemachine.TimeMachineExecutor
import com.rdvdev2.timetravelmod.impl.common.timemachine.TimeMachineManager
import com.rdvdev2.timetravelmod.impl.common.timemachine.exception.TimeMachineExecutionException
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.registry.Registry
import net.minecraft.util.registry.RegistryKey
import net.minecraft.world.Heightmap
import java.util.*

// TODO: Separate packets
object ModNetworking {

    @JvmField
    val OPEN_TIME_MACHINE_GUI = identifier("open_time_machine_gui")

    @JvmField
    val OPEN_CREATIVE_TIME_MACHINE_GUI = identifier("open_creative_time_machine_gui")

    @JvmField
    val OPEN_ENGINEER_BOOK_GUI = identifier("open_engineer_book_gui")

    @JvmField
    val RUN_TIME_MACHINE = identifier("run_time_machine")

    @JvmField
    val RUN_CREATIVE_TIME_MACHINE = identifier("run_creative_time_machine")

    fun register() {
        ServerPlayNetworking.registerGlobalReceiver(RUN_TIME_MACHINE) { server, player, _, buf, _ ->
            val worldIdentifier = buf.readIdentifier()
            val rootPos = buf.readBlockPos()
            val timelineIdentifier = buf.readIdentifier()
            server.execute {
                val world = player.getServer()!!.getWorld(RegistryKey.of(Registry.DIMENSION, worldIdentifier)) ?: return@execute
                val timeline = ModRegistries.TIMELINE[timelineIdentifier] ?: return@execute
                val _tme = TimeMachineManager.generateExecutor(world, rootPos)
                _tme.ifPresent { tme: TimeMachineExecutor ->
                    try {
                        tme.checkAndRun(timeline, player, server)
                    } catch (e: TimeMachineExecutionException) {
                        player.sendMessage(e.error.clientError, true)
                        if (instance.common.enableCheaterReports) {
                            Arrays.stream(server.playerManager.opNames)
                                .map { name: String? -> server.playerManager.getPlayer(name) }
                                .filter { p -> p != player }
                                .forEach { p: ServerPlayerEntity? ->
                                    p?.sendMessage(
                                        e.error.getCheaterReport(
                                            player
                                        ), false
                                    )
                                }
                        }
                    }
                }
            }
        }

        ServerPlayNetworking.registerGlobalReceiver(RUN_CREATIVE_TIME_MACHINE) { server, player, _, buf, _ ->
            val timelineIdentifier = buf.readIdentifier()
            server.execute {
                val timeline = ModRegistries.TIMELINE[timelineIdentifier] ?: return@execute
                val world = server.getWorld(timeline.world)
                if (world == null || world == player.world) return@execute
                player.itemsHand.forEach { itemStack ->
                    if (itemStack.item === ModItems.CREATIVE_TIME_MACHINE) {
                        var pos = world.getTopPosition(Heightmap.Type.MOTION_BLOCKING_NO_LEAVES, player.blockPos)
                        if (pos.y == 0) pos = pos.up(world.seaLevel + 2)
                        while (world.getBlockState(pos).isSolidBlock(world, pos) || world.getBlockState(pos.up()).isSolidBlock(world, pos.up())) pos = pos.up() // Air space
                        while (!world.getBlockState(pos.down()).isSolidBlock(world, pos)) pos = pos.down() // Touching ground
                        // TODO: Special logic for flying players
                        player.teleport(world, player.x, pos.y.toDouble(), player.z, player.yaw, player.pitch)
                    }
                }
            }
        }
    }
}